lib/parsing.js

```diff
@@ -1,5 +1,30 @@
 'use strict';
 
+/**
+ * @typedef AstNode
+ * @property {string} type
+ * @property {'none'|'single'|'double'} quoteStyle?
+ * @property {string} key?
+ * @property {string} name?
+ * @property {string} number?
+ * @property {string} path?
+ * @property {string} string?
+ * @property {boolean} hasEventPrefix?
+ * @property {boolean} typeName?
+ * @property {Object<string,any>} meta?
+ * @property {AstNode} returns?
+ * @property {AstNode} new?
+ * @property {AstNode} value?
+ * @property {AstNode} left?
+ * @property {AstNode} right?
+ * @property {AstNode} owner?
+ * @property {AstNode} subject?
+ * @property {AstNode} this?
+ * @property {AstNode[]} entries?
+ * @property {AstNode[]} objects?
+ * @property {AstNode[]} params?
```

Comment:
I guess it's better than nothing (and better than a few of the alternatives) to have all possible subtype properties merged into a single supertype as optional. If we ever turn on strict null checks, then tsc is going to start complaining though.

If all the contributors are Typescript-aware, and we're running `tsc` as part of tests, then we might as well have a `type.d.ts` file to hold types, and then we can use all the types that TS supports.

Opinions @brettz9 @ExE-Boss ? I was hesitant to do add *any* non-js files because I thought it might impede contributors, but it seems like most people are comfortable around Typescript. 
(I still don't want to convert the code to TS because I like not having a transpile step.)

Reply:
I personally prefer not to convert source to TS, but if you change your mind later, I leave it up to you all. Definition files are fine by me anytime.

Reply:
I agree that a super type is undesirable. It's quite and might even be impossible to do something else in JSDoc, hence why I opted for it.

One problem I find with a `type.d.ts` is that whenever there's a type definition file, then TypeScript as far as I remember switches from looking for JSDoc definitions and instead only uses whatever is in the type definition file.

So one would have to not just add this there, one would need to add everything that's currently defined in JSDoc there.

Reply:
I *think* if you don't use modules in `types.d.ts` then everything works fine. However, I think all the types should go in a definition file at that point, even if the compiler doesn't require it.

Let's leave it as a single supertype inside JSDoc for now. I think it's a good improvement to have SOME type for AstNode instead of nothing.